[Refactor] Add Phase factory methods to `PhaseManager`

innerthunder opened the pull request:
## What are the changes the user will see?

Weather-based form changes from Forecast and Flower Gift should now properly trigger the ability flyout (they previously just constructed `ShowAbilityPhases` without running them).

## Why am I making these changes?

#426 resolved circular imports in phase scheduling in two ways:
- `PhaseId` enum fields were added for each distinct and non-abstract Phase to resolve their types without the use of `instanceof`.
- Bespoke Phase scheduling methods were added to `BattleScene` (now `PhaseManager`) to pull the responsibility of Phase construction from other Phases to the phase manager.

These methods, however, aren't as robust as I and some others would like. Because the basic phase scheduling methods still directly take a phase as input, there's still a risk of circular dependencies popping up as new phases get introduced. The tests in #1236, for example, are currently bricked (likely) because of circular imports between `test` and `src` files.

This PR, modeled after https://github.com/pagefaultgames/pokerogue/pull/5868, improves upon the above methods by:
- Providing alternative Phase scheduling methods that do not require imports
- Adopting Phase identifiers that are more in line with other Phase-related API

## What are the changes from a developer perspective?

- `PhaseManager` (in `src/phase-manager.ts`):
  - Now uses the `PHASES` phase map to construct phases internally
  - Added `createPhase` generic method to construct a Phase from the given constructor parameters
  - Added shorthand methods combining `createPhase` with common scheduling methods:
    - `createAndPushPhase`
    - `createAndUnshiftPhase`
    - `createAndPrependPhase`
    - `createAndAppendPhase`
  - Removed bespoke methods in favor of the new shorthand methods (TODO)
- `Phase` (in `src/phase.ts`):
  - Now uses `phaseName` (a key from `PHASES`) instead of `phaseId` (an enum) as an identifier
  - Rewrote `is` to evaluate phases via the new identifier type

## Screenshots/Videos

<!--
If your changes are changing anything on the user experience, please provide visual proofs of it
Please take screenshots/videos before and after your changes, to show what is brought by this PR
-->

## How to test the changes?

You can playtest this branch at https://poketernity.com/phase-factory/ or run tests locally with `pnpm test:silent`.

## Checklist

- [ ] ⚠️ If this is a PR for `main` (such as a hotfix), has the game version been updated (`pnpm update-version:patch` / `pnpm update-version:minor`?
- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
<!-- We have heavily optimized our test suite, so please actually run the tests :) -->
- [x] Are all unit tests still passing? (`pnpm test:silent`)
  - [ ] Have I created new automated tests (`pnpm test:create`) or updated existing tests related to the PR's changes?
- [ ] Have I provided screenshots/videos of the changes (if applicable)?
  - [ ] Have I made sure that any UI change works for both UI themes (dark and light)?

innerthunder committed:
[Addendum] Add support for multi-Phase input in Phase scheduling methods (#1239)

* Add multi-Phase input for Phase scheduling methods

* Simplify `pushPhase` and `unshiftPhase`

innerthunder committed:
Create `phases.ts`

innerthunder committed:
Add Phase factory utils + update `Phase.is`

innerthunder committed:
Clean up `PhaseManager` utils

innerthunder committed:
Remove `PhaseId`

innerthunder committed:
Replace phase scheduling calls with new `create...` methods

innerthunder committed:
Fix extraneous Phase imports

innerthunder committed:
Add `phaseName` identifier

innerthunder committed:
Update `GameManager.isCurrentPhase`

innerthunder committed:
Remove all XPhase imports in test files (except phaseInterceptor)

innerthunder committed:
Remove low-hanging Phase utils

innerthunder committed:
Remove `queueMessagePhase` (pt 1)

innerthunder committed:
Remove `queueMessagePhase` (pt 2)

innerthunder committed:
Merge branch 'beta' into refactor/no-import-phase-scheduling

innerthunder committed:
Run linters

innerthunder committed:
Fix timing of `ShinySparklePhase` and `PostSummonPhase`

innerthunder committed:
Merge branch 'beta' into refactor/no-import-phase-scheduling

innerthunder committed:
`AFTER_MOVE` battler tags lapse correctly again

innerthunder committed:
Revert accidental `unshift` -> `push` scheduling changes

innerthunder committed:
Fix typo in turn manager

innerthunder committed:
Revert `schedulePhase` shorthand

innerthunder committed:
Fix `PostSummonPhase` timing

innerthunder committed:
Simplify scheduling method params

flx-sta commented on src/phases/stat-stage-change-phase.ts:
I look like a whole lot to me that the `stats` parameter in the constructor could use `coerceArray`. Lots of calls are just 1 stat but always have to be wrapped in an array

innerthunder replied:
Created an issue #1257. This is probably out of this PR's scope.

flx-sta reviewed (commented):
`50` / `371` files viewed

DayKev commented on src/phase-manager.ts:
```diff
@@ -296,110 +541,81 @@ export class PhaseManager {
-   * @param battlerIndex - The user's {@linkcode BattlerIndex}
-   * @param targets - Array of target `BattlerIndex`es
-   * @param move - The {@linkcode PokemonMove} being used
+   * Moves everything from the {@linkcode nextCommandPhaseQueue} to the {@linkcode phaseQueue} (keeping order)
```
```suggestion
   * Queues a new {@linkcode TurnInitPhase}
   * @deprecated We can just queue a new `TurnInitPhase` directly now, to be removed
```

DayKev replied:
Actually this is only used in literally one place within `phase-manager.ts`, just remove it and replace the callsite with the contents of the method.

DayKev commented on src/phase-manager.ts:
```diff
@@ -503,12 +723,10 @@ export class PhaseManager {
    * @param showText - (Optional, default `true`) Whether to show text
```
```suggestion
   * @param showText - (Optional, default `true`) Whether to show text
   * @deprecated To be removed in a future PR
```
Also add to `queueStatStageChangePhase`.

DayKev replied:
Actually there's literally only 2 callsites for `queueStatStageChangePhase` and 3 for `toLoginScreen`, you can just remove them; it won't meaningfully impact the PR size.

DayKev commented on src/phase-manager.ts:
```diff
@@ -73,15 +239,26 @@ interface PokemonFaintInit {
   private phaseQueuePrepend: Phase[] = [];
   /** overrides default of inserting phases to end of phaseQueuePrepend array, useful for inserting Phases "out of order" */
   private phaseQueuePrependSpliceIndex: number = -1;
   private conditionalQueue: [() => boolean, Phase][] = [];
```
```suggestion
  /** @deprecated see {@link https://github.com/Despair-Games/poketernity/pull/910#discussion_r2029764830} */
  private conditionalQueue: [() => boolean, Phase][] = [];
```
Also add to `pushConditionalQueue`.

DayKev commented on src/phase.ts:
```diff
@@ -15,7 +15,7 @@ export abstract class Phase {
 
-  public is<T extends Phase = Phase>(phaseId: T["id"]): this is T {
-    return this.id === phaseId;
+  public is<P extends PhaseKey>(phaseKey: P): this is InstanceType<PhaseMap[P]> {
```
Is `InstanceType<>` supposed to make this act like `instanceof`?

DayKev replied:
Nevermind, I see what happened.

DayKev commented on src/turn-command-manager.ts:
```diff
@@ -214,7 +203,12 @@ export class TurnCommandManager {
   public preemptCommand(commandFilter: TurnCommandFilter): boolean {
     const turnCommand = this.tryRemoveCommand(commandFilter);
-    return !!turnCommand && this.handleCommand(turnCommand);
+    if (!!turnCommand && this.handleCommand(turnCommand)) {
```
```suggestion
    if (turnCommand && this.handleCommand(turnCommand)) {
```

DayKev commented on src/turn-command-manager.ts:
```diff
@@ -363,7 +359,12 @@ export class TurnCommandManager {
   private shiftNextCommand(): boolean {
     const nextCommand = this.turnCommands.shift();
-    return !!nextCommand && this.handleCommand(nextCommand);
+    if (!!nextCommand && this.handleCommand(nextCommand)) {
```
```suggestion
    if (nextCommand && this.handleCommand(nextCommand)) {
```

DayKev commented on src/@types/phase-types.ts:
```diff
@@ -0,0 +1,6 @@
+import type { PhaseMap } from "#app/phase-manager";
+
+export type PhaseKey = keyof PhaseMap;
+export type PhaseConstructorMap = {
```
Why not just use `ConstructorParameters<PhaseMap[P]>` directly (which you're doing for some methods already)?

DayKev replied:
Decided it's fine, updated missed spots to use it (see review PR).

DayKev reviewed (commented):
#1259

DayKev commented on src/data/battler-tags/highest-stat-boost-tag.ts:
```diff
@@ -67,22 +67,20 @@ export abstract class HighestStatBoostTag extends AbilityBattlerTag {
     }
 
-    globalScene.phaseManager.queueMessagePhase(
+    globalScene.phaseManager.createAndUnshiftPhase(
```
```suggestion
    globalScene.phaseManager.createAndPushPhase(
```

DayKev commented on src/data/battler-tags/slow-start-tag.ts:
```diff
@@ -20,12 +20,11 @@ export class SlowStartTag extends AbilityBattlerTag {
     super.onAdd(pokemon);
 
-    globalScene.phaseManager.queueMessagePhase(
+    globalScene.phaseManager.createAndUnshiftPhase(
```
```suggestion
    globalScene.phaseManager.createAndPushPhase(
```

DayKev commented on src/data/mystery-encounters/encounters/fun-and-games-encounter.ts:
```diff
@@ -395,13 +393,13 @@ function summonPlayerPokemonAnimation(pokemon: PlayerPokemon): Promise<void> {
 
                   globalScene.triggerPokemonFormChange(pokemon, SpeciesFormChangeActiveTrigger, true);
-                  globalScene.phaseManager.pushPhase(new PostSummonPhase(pokemon.getBattlerIndex()));
+                  globalScene.phaseManager.createAndUnshiftPhase("PostSummonPhase", pokemon.getBattlerIndex());
```
```suggestion
                  globalScene.phaseManager.createAndPushPhase("PostSummonPhase", pokemon.getBattlerIndex());
```

DayKev commented on src/field/arena.ts:
```diff
@@ -481,7 +479,8 @@ export class Arena {
 
       if (isCastformWithForecast || isCherrimWithFlowerGift) {
-        new ShowAbilityPhase(p.getBattlerIndex());
+        /** @todo This doesn't seem to account for which ability is triggered (main vs. passive) */
```
```suggestion
        // TODO: This doesn't seem to account for which ability is triggered (main vs. passive)
```

DayKev commented on src/field/arena.ts:
```diff
@@ -498,7 +497,8 @@ export class Arena {
 
       if (isCastformWithForecast || isCherrimWithFlowerGift) {
-        new ShowAbilityPhase(p.getBattlerIndex());
+        /** @todo This doesn't seem to account for which ability is triggered (main vs. passive) */
```
```suggestion
        // TODO: This doesn't seem to account for which ability is triggered (main vs. passive)
```

DayKev commented on src/phases/abstract-hit-check-phase.ts:
It was decided we shouldn't put `abstract-` in filenames, so this file shouldn't get renamed.

innerthunder replied:
I just renamed and reorganized all of the abstract Phase files into a `base` subfolder

DayKev commented on src/phases/encounter-phase.ts:
```diff
@@ -550,13 +541,13 @@ export class EncounterPhase extends BattlePhase {
     enemyField.forEach((enemyPokemon, e) => {
       if (enemyPokemon.isShiny()) {
-        globalScene.phaseManager.unshiftPhase(new ShinySparklePhase(BattlerIndex.ENEMY + e));
+        phaseManager.createAndPushPhase("ShinySparklePhase", BattlerIndex.ENEMY + e);
```
```suggestion
        phaseManager.createAndUnshiftPhase("ShinySparklePhase", BattlerIndex.ENEMY + e);
```

DayKev commented on src/phases/faint-phase.ts:
```diff
@@ -207,23 +203,28 @@ export class FaintPhase extends PokemonPhase {
         if (hasReservePartyMember) {
-          globalScene.phaseManager.pushPhase(
-            new SwitchSummonPhase(SwitchType.SWITCH, this.fieldIndex, -1, false, false),
+          globalScene.phaseManager.createAndUnshiftPhase(
```
```suggestion
          globalScene.phaseManager.createAndPushPhase(
```

DayKev commented on test/abilities/imposter.test.ts:
```diff
@@ -127,7 +126,7 @@ describe("Abilities - Imposter", () => {
     await game.toNextWave();
 
-    expect(game.scene.phaseManager.getCurrentPhase()?.id).toBe(PhaseId.COMMAND);
+    expect(game.scene.phaseManager.getCurrentPhase()?.is("CommandPhase")).toBeTruthy();
```
```suggestion
    expect(game.scene.phaseManager.getCurrentPhase()?.phaseName).toBe("CommandPhase");
```

DayKev replied:
This way the error message if the test fails will tell us what the actual phase was, instead of just "expected `true` got `false`".

DayKev commented on test/battler-tags/octolock.test.ts:
```diff
@@ -31,7 +31,7 @@ describe("BattlerTag - OctolockTag", () => {
 
       vi.spyOn(game.scene.phaseManager, "unshiftPhase").mockImplementation((phase) => {
-        expect(phase).toBeInstanceOf(StatStageChangePhase);
+        expect(phase.is("StatStageChangePhase")).toBeTruthy();
```
```suggestion
        expect(phase.phaseName).toBe("StatStageChangePhase");
```

DayKev commented on test/moves/focus-punch.test.ts:
```diff
@@ -106,10 +105,10 @@ describe("Moves - Focus Punch", () => {
     await game.phaseInterceptor.to("TurnStartPhase");
 
-    expect(game.scene.phaseManager.getCurrentPhase()).toMatchObject({ id: PhaseId.SWITCH_SUMMON });
+    expect(game.scene.phaseManager.getCurrentPhase()?.is("SwitchSummonPhase")).toBeTruthy();
```
```suggestion
    expect(game.scene.phaseManager.getCurrentPhase()?.phaseName).toBe("SwitchSummonPhase");
```

DayKev commented on test/moves/focus-punch.test.ts:
```diff
@@ -106,10 +105,10 @@ describe("Moves - Focus Punch", () => {
     await game.phaseInterceptor.to("PostActionPhase");
 
-    expect(game.scene.phaseManager.getCurrentPhase()).toMatchObject({ id: PhaseId.MOVE_HEADER });
+    expect(game.scene.phaseManager.getCurrentPhase()?.is("MoveHeaderPhase")).toBeTruthy();
```
```suggestion
    expect(game.scene.phaseManager.getCurrentPhase()?.phaseName).toBe("MoveHeaderPhase");
```

DayKev commented on test/moves/shell-trap.test.ts:
```diff
@@ -50,7 +49,7 @@ describe("Moves - Shell Trap", () => {
 
     const movePhase = game.scene.phaseManager.getCurrentPhase();
-    expect(movePhase?.id).toBe(PhaseId.MOVE);
+    expect(movePhase?.is("MovePhase")).toBeTruthy();
```
```suggestion
    expect(movePhase?.phaseName).toBe("MovePhase");
```

DayKev commented on test/moves/shell-trap.test.ts:
```diff
@@ -73,7 +72,7 @@ describe("Moves - Shell Trap", () => {
 
     const movePhase = game.scene.phaseManager.getCurrentPhase();
-    expect(movePhase?.is<MovePhase>(PhaseId.MOVE)).toBeTruthy();
+    expect(movePhase?.is("MovePhase")).toBeTruthy();
```
```suggestion
    expect(movePhase?.phaseName).toBe("MovePhase");
```

DayKev commented on test/moves/shell-trap.test.ts:
```diff
@@ -96,7 +95,7 @@ describe("Moves - Shell Trap", () => {
 
     const movePhase = game.scene.phaseManager.getCurrentPhase();
-    expect(movePhase?.is<MovePhase>(PhaseId.MOVE)).toBeTruthy();
+    expect(movePhase?.is("MovePhase")).toBeTruthy();
```
```suggestion
    expect(movePhase?.phaseName).toBe("MovePhase");
```

DayKev commented on test/moves/shell-trap.test.ts:
```diff
@@ -117,7 +116,7 @@ describe("Moves - Shell Trap", () => {
 
     const movePhase = game.scene.phaseManager.getCurrentPhase();
-    expect(movePhase?.is<MovePhase>(PhaseId.MOVE)).toBeTruthy();
+    expect(movePhase?.is("MovePhase")).toBeTruthy();
```
```suggestion
    expect(movePhase?.phaseName).toBe("MovePhase");
```

DayKev commented on test/moves/transform.test.ts:
```diff
@@ -133,7 +132,7 @@ describe("Moves - Transform", () => {
     await game.toNextWave();
 
-    expect(game.scene.phaseManager.getCurrentPhase()?.id).toBe(PhaseId.COMMAND);
+    expect(game.scene.phaseManager.getCurrentPhase()?.is("CommandPhase")).toBeTruthy();
```
```suggestion
    expect(game.scene.phaseManager.getCurrentPhase()?.phaseName).toBe("CommandPhase");
```

DayKev commented on test/mystery-encounter/encounters/a-trainers-test-encounter.test.ts:
```diff
@@ -107,7 +104,7 @@ describe("A Trainer's Test - Mystery Encounter", () => {
 
       const enemyField = scene.getEnemyField();
-      expect(scene.phaseManager.getCurrentPhase()?.constructor.name).toBe(CommandPhase.name);
+      expect(game.isCurrentPhase("CommandPhase")).toBeTruthy();
```
```suggestion
      expect(scene.phaseManager.getCurrentPhase()?.phaseName).toBe("CommandPhase");
```

innerthunder replied:
Do we even want the `isCurrentPhase` helper if this is the preferred assertion?

DayKev replied:
Well `isCurrentPhase` wasn't made to be used in `expect`s.

DayKev replied:
You could also use a custom error message (or a custom matcher, like `expect(game.scene.phaseManager.getCurrentPhase()).toBePhase("CommandPhase")`).

DayKev replied:
It might make sense to make a `GameManager#currentPhase` helper as well.

DayKev commented on test/mystery-encounter/encounters/a-trainers-test-encounter.test.ts:
```diff
@@ -126,7 +123,7 @@ describe("A Trainer's Test - Mystery Encounter", () => {
       await skipBattleRunMysteryEncounterRewardsPhase(game);
       await game.phaseInterceptor.to("SelectModifierPhase", false);
-      expect(scene.phaseManager.getCurrentPhase()?.constructor.name).toBe(SelectModifierPhase.name);
+      expect(game.isCurrentPhase("SelectModifierPhase")).toBeTruthy();
```
```suggestion
      expect(scene.phaseManager.getCurrentPhase()?.phaseName).toBe("SelectModifierPhase");
```

DayKev commented on test/mystery-encounter/encounters/a-trainers-test-encounter.test.ts:
```diff
@@ -174,7 +171,7 @@ describe("A Trainer's Test - Mystery Encounter", () => {
       await runMysteryEncounterToEnd(game, 2);
       await game.phaseInterceptor.to("SelectModifierPhase", false);
-      expect(scene.phaseManager.getCurrentPhase()?.constructor.name).toBe(SelectModifierPhase.name);
+      expect(game.isCurrentPhase("SelectModifierPhase")).toBeTruthy();
```
```suggestion
      expect(scene.phaseManager.getCurrentPhase()?.phaseName).toBe("SelectModifierPhase");
```

DayKev replied:
etc etc

DayKev committed:
Review for PR 1250 (#1259)

Update types, implement in `phase-interceptor.ts`

This restores type safety for `PhaseInterceptor#to`

innerthunder committed:
Apply review suggestions

innerthunder committed:
Reorganize abstract phases into sub-folder

innerthunder committed:
Replace `isCurrentPhase` assertions

innerthunder committed:
Merge branch 'beta' into refactor/no-import-phase-scheduling

innerthunder committed:
Merge branch 'beta' into refactor/no-import-phase-scheduling

flx-sta commented on src/data/battler-tags/slow-start-tag.ts:
```diff
@@ -19,12 +19,11 @@ export class SlowStartTag extends AbilityBattlerTag {
+      undefined,
       false,
-      null,
-      true,
```
Is this `true` value no longer needed?

DayKev replied:
Here `true` meant "use `pushPhase`".

flx-sta commented on src/data/berry.ts:
```diff
@@ -101,8 +102,8 @@ export function getBerryEffectFunc(berryType: BerryType): BerryEffectFunc {
         const statStages = new NumberHolder(1);
         applyAbAttrs<DoubleBerryEffectAbAttr>(AbAttrFlag.DOUBLE_BERRY_EFFECT, pokemon, false, statStages);
-        globalScene.phaseManager.queueStatStageChangePhase(
-          true,
+        globalScene.phaseManager.createAndUnshiftPhase(
+          "StatStageChangePhase",
           pokemon.getBattlerIndex(),
```
What happened to the `true` param?

DayKev replied:
`true` meant "use `unshiftPhase`" for the removed function.

flx-sta commented on src/data/berry.ts:
```diff
@@ -122,8 +123,8 @@ export function getBerryEffectFunc(berryType: BerryType): BerryEffectFunc {
         const stages = new NumberHolder(2);
         applyAbAttrs<DoubleBerryEffectAbAttr>(AbAttrFlag.DOUBLE_BERRY_EFFECT, pokemon, false, stages);
-        globalScene.phaseManager.queueStatStageChangePhase(
-          true,
+        globalScene.phaseManager.createAndUnshiftPhase(
+          "StatStageChangePhase",
           pokemon.getBattlerIndex(),
```
What happened to the `true` param?

DayKev replied:
Same here.

flx-sta reviewed (commented):
Halftime `188`/`376`

DayKev commented on src/phases/command-phase.ts:
```diff
@@ -392,8 +391,7 @@ export class CommandPhase extends FieldPhase {
     if (this.fieldIndex) {
-      globalScene.phaseManager.unshiftPhase(new CommandPhase(0));
-      globalScene.phaseManager.unshiftPhase(new CommandPhase(1));
+      globalScene.phaseManager.pushPhase(new CommandPhase(0), new CommandPhase(1));
```
Oh I missed this the first time.
```suggestion
      globalScene.phaseManager.unshiftPhase(new CommandPhase(0), new CommandPhase(1));
```

DayKev commented on src/phase-manager.ts:
```diff
@@ -524,11 +716,10 @@ export class PhaseManager {
     eager: boolean,
     ...params: ConstructorParameters<typeof StatStageChangePhase>
   ): void {
```
Forgot to remove this method.

innerthunder committed:
Review suggestions + Remove `queueShowAbility`

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

DayKev commented on src/utils/ability-utils.ts:
```diff
@@ -23,11 +22,6 @@ export function getPokemonWithWeatherBasedForms(): Pokemon[] {
     );
 }
 
-export function queueShowAbility(pokemon: Pokemon, passive: boolean): void {
-  globalScene.phaseManager.unshiftPhase(new ShowAbilityPhase(pokemon.id, passive));
-  globalScene.phaseManager.clearPhaseQueueSplice();
-}
```
Wow I didn't even know this was here lol.

innerthunder committed:
Merge branch 'beta' into refactor/no-import-phase-scheduling

innerthunder committed:
Merge branch 'beta' into refactor/no-import-phase-scheduling

DayKev committed:
Merge branch 'beta' into refactor/no-import-phase-scheduling

DayKev committed:
Merge branch 'beta' into refactor/no-import-phase-scheduling

DayKev committed:
Merge branch 'beta' into refactor/no-import-phase-scheduling

innerthunder committed:
Merge branch 'beta' into refactor/no-import-phase-scheduling

flx-sta commented:
@innerthunder Unfortunately there is conflicts to resolve, otherwise lgtm

innerthunder committed:
Merge branch 'beta' into refactor/no-import-phase-scheduling
